### PR TITLE
Add tasklist and task class, with importance enum

### PR DESCRIPTION
**Pull request by wish2023**

### CODE REVIEW

**Review by daniellimws — APPROVED** [review]
Looks good! But I think it would be nice to standardize the number of empty lines between methods. Sometimes there are only 1, and sometimes there are 2 or 3. Let's stick with 1 empty line between methods?

File: src/main/java/athena/TaskList.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+
+    /**
+     * Marks specified task as done.
+     * Lets the user know specified task has been marked as done.
+     *
+     * @param task Task to be marked
+     */
+    public void markTaskAsDone(String task) {
+        int taskNumber = getTaskNumber(task);
```
Comment (amanda-chua) [inline src/main/java/athena/TaskList.java]: For this i will be parsing in an integer alr so i think can remove the getTaskNumber method.

File: src/main/java/athena/TaskList.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+
+    /**
+     * Deletes the task at the specified position in the task list.
+     *
+     * @param index Position of task in task list
+     */
+    public void deleteTask(String index) {
+        int taskNumber = getTaskNumber(index);
```
Comment (amanda-chua) [inline src/main/java/athena/TaskList.java]: For this i will be parsing in an integer alr so i think can remove the getTaskNumber method.

File: src/main/java/athena/TaskList.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
+package athena;
```
Comment (amanda-chua) [inline src/main/java/athena/TaskList.java:1]: Can replace all the println to UI methods.

**Review by amanda-chua — CHANGES_REQUESTED** [review]
Looks good!

File: src/main/java/athena/TaskList.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * Deletes the task at the specified position in the task list.
+     *
+     * @param taskNumber Position of task in task list
+     */
+    public void deleteTask(int taskNumber) {
+        tasks.remove(taskNumber);
+    }
+
```
Comment (amanda-chua) [inline src/main/java/athena/TaskList.java:104]: Remove one empty line from here.

File: src/main/java/athena/TaskList.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public void displayList(String importanceFilter) {
+        if (importanceFilter.equals(NO_FILTER)) {
+            printAllTasks();
+        } else {
+            printTasksWithFilter(importanceFilter);
+        }
+    }
+
+
```
Comment (amanda-chua) [inline src/main/java/athena/TaskList.java]: Remove empty lines from here.

File: src/main/java/athena/TaskList.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    private void printTasksWithFilter(String importanceFilter) {
+        for (Task task : tasks) {
+            if (task.getImportance().equals(importanceFilter)) {
+                System.out.println(task);
+            }
+        }
+    }
+
+
```
Comment (amanda-chua) [inline src/main/java/athena/TaskList.java]: Remove one empty line from here.

File: src/main/java/athena/Task.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Marks the task as done.
+     *
+     */
+    public void setDone() {
+        isDone = true;
+    }
+
+
```
Comment (amanda-chua) [inline src/main/java/athena/Task.java]: Remove one empty line from here.

File: src/main/java/athena/Task.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        this.name = name;
+        this.startTime = startTime;
+        this.duration = duration;
+        this.deadline = deadline;
+        this.recurrence = recurrence;
+        this.importance = importance;
+        this.notes = notes;
+    }
+
+
```
Comment (amanda-chua) [inline src/main/java/athena/Task.java]: Remove one empty line from here.

**Review by amanda-chua — APPROVED** [review]
Good!

File: src/main/java/athena/TaskList.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        for (Task task: tasks) {
+            if (task.getName().contains(instance)) {
+                shortlistedTasks.add(task);
+            }
+        }
+        return shortlistedTasks;
+    }
+
+    private void printAllTasks() {
+        for (Task task : tasks) {
```
Comment (alstontham) [inline src/main/java/athena/TaskList.java]: Can call "ui.printList(ArrayList<Task> tasks)" and let the ui handle the printing

File: src/main/java/athena/TaskList.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            System.out.println(task);
+        }
+    }
+
+    private void printTasksWithFilter(String importanceFilter) {
```
Comment (alstontham) [inline src/main/java/athena/TaskList.java]: Can call the printList() function from ui and pass it the filtered list (so tasklist will still handle the filtering)

File: src/main/java/athena/TaskList.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+     * @param taskNumber Index of task
+     * @param name Name of task
+     * @param startTime Start time of task
+     * @param duration Duration of task
+     * @param deadline Deadline of task
+     * @param recurrence Recurrence of task
+     * @param importance Importance of task
+     * @param notes Additional notes of task
+     */
+    public void editTask(int taskNumber, String name, String startTime, String duration,
```
Comment (alstontham) [inline src/main/java/athena/TaskList.java:117]: Could call ui method printTaskEdited(task) to print out the edited task to the user

File: src/main/java/athena/TaskList.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        return tasks.get(index).toString();
+    }
+
+    /**
+     * Deletes the task at the specified position in the task list.
+     *
+     * @param taskNumber Position of task in task list
+     */
+    public void deleteTask(int taskNumber) {
+        tasks.remove(taskNumber);
```
Comment (alstontham) [inline src/main/java/athena/TaskList.java:101]: Could call ui method printTaskDeleted(task) to print out the deleted task here as well

File: src/main/java/athena/TaskList.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+     * @param duration Duration of task
+     * @param deadline Deadline of task
+     * @param recurrence Recurrence of task
+     * @param importance Importance of task
+     * @param notes Additional notes of task
+     */
+    public void addToList(String name, String startTime, String duration,
+                String deadline, String recurrence, String importance, String notes) {
+        Task task = createTask(name, startTime, duration, deadline, recurrence, importance, notes);
+        tasks.add(task);
```
Comment (alstontham) [inline src/main/java/athena/TaskList.java:80]: Can call ui method printTaskAdded(task) to let users know what task they added

File: src/main/java/athena/TaskList.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    /**
+     * Marks specified task as done.
+     * Lets the user know specified task has been marked as done.
+     *
+     * @param taskNumber Position of task in task list
+     */
+    public void markTaskAsDone(int taskNumber) {
+        tasks.get(taskNumber).setDone();
```
Comment (alstontham) [inline src/main/java/athena/TaskList.java:61]: Can call the ui method printTaskDone(task) to let users know which task they marked as done

**Review by alstontham — CHANGES_REQUESTED** [review]
Looks good! If you call the ui methods at the points I commented then we can let the ui class deal with all of the printing out to the user!

File: src/main/java/athena/Task.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+     *
+     * @param name New task name
+     * @param startTime New task start time
+     * @param duration New task duration
+     * @param deadline New task deadline
+     * @param recurrence New task recurrence
+     * @param importance New task importance
+     * @param notes New task notes
+     */
+    public void changeAttributesTo(String name, String startTime, String duration,
```
Comment (daniellimws) [inline src/main/java/athena/Task.java]: The `To` at the end of the method name feels unnecessary. Shall we use `changeAttributes`? Or `editAttributes` or `edit` which I think will match the user functionality more closely too.

File: src/main/java/athena/TaskList.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+     *
+     * @param name Name of task
+     * @param startTime Start time of task
+     * @param duration Duration of task
+     * @param deadline Deadline of task
+     * @param recurrence Recurrence of task
+     * @param importance Importance of task
+     * @param notes Additional notes of task
+     */
+    public void addToList(String name, String startTime, String duration,
```
Comment (daniellimws) [inline src/main/java/athena/TaskList.java]: I think naming this method as `addTask` would be better as it is consistent with the other method names such as `editTask` and `deleteTask`.

File: src/main/java/athena/TaskList.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+        tasks.add(task);
+    }
+
+    /**
+     * Returns the task description at the specified position in task list.
+     *
+     * @param index Position of task in the task list
+     * @return Task description
+     */
+    public String getLine(int index) {
```
Comment (daniellimws) [inline src/main/java/athena/TaskList.java]: I think `getTaskDescription` would be a more descriptive name for this method?

**Review by daniellimws — COMMENTED** [review]
Looks good! Just left some comments about the method names.

**Comment by daniellimws** [comment]
Can you add a method to get the whole list without any filtering?

File: src/main/java/athena/Task.java (hunk 15 above)
Comment (wish2023) [reply src/main/java/athena/Task.java]: Noted! Has been changed to `edit`

File: src/main/java/athena/TaskList.java (hunk 16 above)
Comment (wish2023) [reply src/main/java/athena/TaskList.java]: @amanda-chua take note method name has been changed to `addTask`

**Pull request merged**